src/core/lwip/tcp_out.c

```diff
@@ -1885,6 +1885,44 @@ void tcp_rexmit_rto(struct tcp_pcb *pcb)
     if (pcb->unacked == NULL) {
         return;
     }
+    /*  Before moving all unacked to unsent, we need to maintain correct order of segments.
+        For exmaple, in unsent we might find a retransmitted segments:
+        unacked: 1->2->3
+        unsent: 4->5->6
+
+        After fast retransmission of seg 1,2:
+        unacked: 3
+        unsent: 1->2->4->5->6
+
+        After RTO without handling correct order:
+        unacked: ----
+        unsent: 3->1->2->4->5->6
+
+        Therefore - we first move back retransmitted segments to unacked:
+        unacked: 1->2->3
+        unsent: 4->5->6
+        And then move all unacked to unsent:
+        unacked: ----
+        unsent: 1->2->3->4->5->6
+    */
+
+    if (pcb->unsent != NULL && TCP_SEQ_GT(pcb->unacked->seqno, pcb->unsent->seqno)) {
+        // Move retransmitted segments to unacked
+        struct tcp_seg *orig_unsent = pcb->unsent;
+        struct tcp_seg *last_unsent_move_to_unacked = pcb->unsent;
+        while (last_unsent_move_to_unacked->next != NULL &&
+               TCP_SEQ_GT(pcb->unacked->seqno, last_unsent_move_to_unacked->next->seqno)) {
+            last_unsent_move_to_unacked = last_unsent_move_to_unacked->next;
+        }
+
+        pcb->unsent = last_unsent_move_to_unacked->next;
+        if (pcb->unsent == NULL) {
+            pcb->last_unsent = NULL;
+        }
+
+        last_unsent_move_to_unacked->next = pcb->unacked;
+        pcb->unacked = orig_unsent;
+    }
```

Comment:
Is it guaranteed that the retransmitted segments in the unsent are in order? If there is a way to call fast retransmission twice, the order would be reversed. So, either each retransmitted segment needs to be inserted to the right place individually or only the single head element needs to be moved.

Reply:
It's working properly.